coding/python_testing.rst

```diff
@@ -4,6 +4,156 @@ Python Unit Testing
 
 This page provides technical guidance to developers writing unit tests for DM's Python code base.
 See :doc:`unit_test_policy` for an overview of LSST Stack testing.
+If you have legacy DM :mod:`unittest` ``suite``-based code (code that sets up a :class:`unittest.TestSuite` object by listing specific test classes and that uses :func:`lsst.utils.tests.run` rather than :func:`unittest.main`) please refer to tech note `SQR-012`_ for porting instructions.
+LSST tests should be written using the :mod:`unittest` framework, with default test discovery, and should support being run using the `pytest`_ test runner.
+
+.. _SQR-012: http://sqr-012.lsst.io
+.. _pytest: http://pytest.org
+
+Introduction to ``unittest``
+============================
+
+This document will not attempt to explain full details of how to use :mod:`unittest` but instead shows common scenarios encountered in the LSST codebase.
+
+A simple :mod:`unittest` example is shown below:
+
+.. literalinclude:: unit_test_snippets/unittest_basic_example.py
+   :language: python
+
+The important things to note in this example are:
+
+* If the test is being executed using :command:`python` from the command line the :py:func:`unittest.main` call performs the test discovery and executes the tests, setting exit status to non-zero if any of the tests fail.
+* Test classes are executed in the order in which they appear in the test file.
+  In this case the tests in :class:`DemoTestCase1` will be executed before those in :class:`DemoTestCase2`.
+* Test classes must, ultimately, inherit from :class:`unittest.TestCase` in order to be discovered, and tests themselves must be methods that start with ``test``.
+  All other methods and classes will be ignored by the test system but can be used by tests.
+* If a test method completes the test passes, if it throws an uncaught exception the test has failed.
+
+Using a more advanced test runner
+=================================
+
+All tests should be written such that they are runnable using `pytest`_, a policy adopted in `RFC-69`_.
+`pytest`_ provides a much richer execution and reporting environment for tests and can be used to run multiple tests files together.
+`pytest`_ test discovery is much more flexible than that provided by :mod:`unittest` but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner.
+In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
+
```

Comment:
I don't think pytest can be configured to solely use unittest default test discovery. It will be somewhat obvious when the person writes the tests so I don't think it's a problem once we have a stack that works with pytest. There is going to be work required to transition our test suites.